hnn_core/gui/gui.py

```diff
@@ -1118,9 +1130,41 @@ def load_drive_and_connectivity(params, log_out, drives_out,
                       layout)
 
 
-def on_upload_change(change, params, tstop, dt, log_out, drive_boxes,
-                     drive_widgets, drives_out, connectivity_out,
-                     connectivity_textfields, layout, load_type):
+def on_upload_data_change(change, data, viz_manager):
+    if len(change['owner'].value) == 0:
+        logger.info("Empty change")
+        return
+
+    key = list(change['new'].keys())[0]
+
+    data_fname = change['new'][key]['metadata']['name'].rstrip('.txt')
+    if data_fname in data['simulation_data'].keys():
+        logger.warn(f"Overwrite the existing data {data_fname}")
+
+    ext_content = change['new'][key]['content']
+    ext_content = codecs.decode(ext_content, encoding="utf-8")
+    e_c = io.StringIO(ext_content)
+    dpls = list()
+    dpl_data = np.loadtxt(e_c, dtype=float)
+    for idx in range(1, dpl_data.shape[1]):
+        dpls.append(hnn_core.Dipole(dpl_data[:, 0], dpl_data[:, idx]))
```

Comment:
we should use `read_dipole` as much as possible. Could `read_dipole` be modified to also accept a file object?

Reply:
It can do that without any modification. Just passing the `StringIO` object will be fine. The problem is that `read_dipole` will drop arrays whose 2nd dimension is neither 1 nor 3.